### PR TITLE
Support overriding ROWCOUNT for SQL Server

### LINKED ISSUE #9940: ROWCOUNT limited ON SqlServer
System : Win 10，Java 1.8
Data On SQL Server 2008
The results is affected by ROWCOUNT limited when use "Native Query",if the Metabase version >0.31.0.

```
Test SQL script 
----------------------------------------
DECLARE @DATA AS TABLE(
    IDX INT IDENTITY(1,1),
    V INT
)
DECLARE @STEP INT 
SET @STEP = 1
WHILE @STEP <=3000
BEGIN
    INSERT INTO @DATA(V)
    SELECT 1
    SET @STEP = @STEP + 1
END 

DECLARE @TEMP AS TABLE(
    IDX INT IDENTITY(1,1),
    V INT
)
INSERT INTO @TEMP(V)
SELECT V FROM @DATA

SELECT COUNT(1) FROM @TEMP
---------------------------------------

```
The data has 3000 rows,maybe more,but the result is limited at 2000.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by jeff303**
Add new driver multimethod called `row-limit-override` which, given a `driver` and `database`, can optionally return a row-limit to be applied to a query (instead of the default value)

Implement that multimethod for :sqlserver by looking up a new (optional) connection property named `:rowcount-override`

Marking `sql-jdbc.execute/statement-supported?` as `false` for `:sqlserver` (since there is an error when executing native queries like the type in #9940 using a `Statement`); the driver will always use `PreparedStatement` now

Move common computation function of overall row limit to the limit namespace, and invoke it from execute now, called `determine-query-max-rows`

Add new clause to the `determine-query-max-rows` function that preferentially takes the value from `row-limit-override` (if defined)

Adding new test case to confirm that `:sqlserver` takes the new row limit when executing a native query that has intermediate temp results

### CODE REVIEW

**Comment by jeff303** [comment]
This was the best way I could figure on how to thread the needle on #9940

- A new optional connection property for SQL server that allows for overriding the `ROWCOUNT` parameter (by way of `.setMaxRows` on the `PreparedStatement`)
- A new multimethod that allows overriding the DB-specific row limit (which in the case of SQL server, returns that value if defined)

I don't love having to add a new multimethod for this, but couldn't think of a cleaner alternative.

File: src/metabase/query_processor/middleware/limit.clj [hunk 1]
```diff
@@ -27,13 +30,26 @@
            (ensure-reduced result')
            result'))))))
 
+(defn determine-query-max-rows
```
Comment (howonlee) [inline src/metabase/query_processor/middleware/limit.clj:33]: there's a pretty similar thing in `metabase.query-processor.middleware.add-rows-truncated` which should prolly be folded in

**Review by howonlee — APPROVED** [review]
LGTM but I note there's a thingy in `add-rows-truncated` that purports to do a similar thing as `determine-query-max-rows` and it may be worth doublechecking for an interaction with respect to the limit we splice in strangely in sql server order by statements in the multi `sql.qp/apply-top-level-clause [:sqlserver :order-by]`

**Review by camsaul — CHANGES_REQUESTED** [review]
I prefer what you proposed in https://github.com/metabase/metabase/issues/9940#issuecomment-958165654 (turning `max-results-bare-rows` into a Setting) over making this a per-DB thing. We don't need any more knobs when connecting a database -- the big selling point of Metabase is that it's easy to set up, and every new knob we add to DB details is taking away from that. 

I would just make this an `:internal` Setting so people can set this via env var and at least there's a workaround for the issues that people are running in to (until we come up with a better solution -- still not sure what that is). I'd vote to make this env-var-only because we also don't want Yet Another Knob in the admin panel. The basic rule is do not add any more knobs unless you absolutely have a really good reason to, and I'm not convinced we do here (if we don't know what the appropriate value for this Setting is, how are we going to document this to tell people what they should put here?) 

Ideally, we'd just figure out what number to use for SQL Server queries (maybe it actually is `0`, or at least something higher than `2000`) and set that so Metabase does the right thing out of the box. That's the [Zen of Metabase](https://github.com/metabase/metabase/blob/master/zen.md) solution at least -- we shouldn't _require_ people to guess some number and set it in the admin panel for things to work sensibly.

That's my 2 cents.

Backing up a bit, if we decide we do want to make this configurable on a per-Database basis, I would suggest not introducing any new methods. The other rule is the driver interface is already complicated enough and we should not make it any more so unless we have a really good reason...

You could just pick some generic DB details key name like `:row-limit` (maybe namespace it as `:metabase/row-limit` so we don't confuse it with connection properties?) and just have `determine-query-max-rows` look for that property and return it if it exists. (Do we really want to have to introduce nearly identical connection properties and method implementations for every other Database we support if we decide we want people to let people set this for Postgres/MySQL/etc. in the future?) If you do it this way you avoid making the driver API harrier and you get this functionality for free everywhere (even if not exposed for the other DBs in the UI, but people could still theoretically set this property by hand).

I think this would actually be an appropriate use case for Database-local Settings (#14055) _if_ we already had them. We could make `max-results-bare-rows` a Setting and allow it to be overridden on a per-Database basis. We don't have that yet, but it's still useful to think of this PR as conceptually equivalent to a Database-local Setting.

At any rate, I don't think we want to have to introduce a new multimethod and some one-off arbitrary keys for every single thing we want to be configurable per DB. It's not a pattern that's going to scale well in the long run...

TL;DR

1. Most of all I'd prefer just figuring out how to fix #9940 without introducing Yet Another Knob
2. Failing that, I'd prefer if we just went with the simplest solution and made the hardcoded max results limit of 2000 configurable via env var
3. Failing that, let's just bite the bullet and implement #14055 so we have a general way to approach this sort of thing in the future and don't have to implement a bunch of one-off methods and connection properties the next time we want to make some random constant configurable. We wanted #14055 anyway for other stuff and it's not like it would be that hard to do

**Comment by jeff303** [comment]
> TL;DR
> 
> 1. Most of all I'd prefer just figuring out how to fix #9940 without introducing Yet Another Knob
> 2. Failing that, I'd prefer if we just went with the simplest solution and made the hardcoded max results limit of 2000 configurable via env var
> 3. Failing that, let's just bite the bullet and implement #14055 so we have a general way to approach this sort of thing in the future and don't have to implement a bunch of one-off methods and connection properties the next time we want to make some random constant configurable. We wanted #14055 anyway for other stuff and it's not like it would be that hard to do

Yeah, I wish there was some way to tell a priori whether the override is even needed, on a per query basis, without requiring any sort of configuration.  For instance, we know this will only ever be an issue on native queries (and really, only on ones similar to the description in #9940, not that we want to try to parse that craziness), so we could at least restrict the behavior there.

I agree that a new driver multimethod is overkill here (see comment above).  My rationale for wanting to keep the override as part of the specific DB instance was that certain admins may want to allow it in some cases, but not others, but that could obviously be accomplished via a Database-local setting.  I think that shoehorning into an instance level setting would feel like an anit-pattern.  My inclination is therefore to not do this until #14055 can be done.

**Comment by camsaul** [comment]
I'm taking a crack at getting Database-local Settings working really quick

**Comment by jeff303** [comment]
@camsaul , this is updated to use the DB local setting (I believe).  Mind having a look?

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj [hunk 2]
```diff
@@ -447,6 +447,26 @@
...
+  (let [stmt (.prepareStatement conn sql
+                                ResultSet/TYPE_FORWARD_ONLY
+                                ResultSet/CONCUR_READ_ONLY)]
+    (try
+      (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+      (sql-jdbc.execute/set-parameters! driver stmt params)
+      stmt
+      (catch Throwable e
+        (.close stmt)
+        (throw e)))))
```
Comment (camsaul) [inline modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj:468]: looks like this will have no effect since the method is defined again at https://github.com/metabase/metabase/blob/f47daacffc3e446ea41a9ea8b56309d54fbd73c8/modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj#L535-L550

**Review by camsaul — APPROVED** [review]
👍 this works for me but please fix the duplicate impls of `prepared-statement` I mentioned before you merge it. (it seems like it probably wasn't needed as part of this fix anyway since it's dead code?)

File: modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj (hunk 2 above)
Comment (jeff303) [reply modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj:473]: Fixed.

**Comment by rlotun** [comment]
is there any way we can get this into `release-x.42.x` (since this is a P1 bugfix)?

**Comment by rlotun** [comment]
> is there any way we can get this into `release-x.42.x` (since this is a P1 bugfix)?

ah nevermind, it's fine - I see it was communicated as being in 0.43. no worries.
